### ISSUE TITLE
[Bug] ArrayDataProvider::getKeys()

### ISSUE THREAD

**Issue by mdmunir** [yiisoft/yii2#8073]
`ArrayDataProvider::getKeys()` return wrong if `ArrayDataProvider::$allModels` has integer key.
```php
$items = [
    'a' => [...],
    'b' => [...],
    'c' => [...],
    10 => [...], // integer key
    20 => [...],
];

$provider = new ArrayDataProvider(['allModels'=>$items]);
VarDumper::dump($provider->getKeys());

// output are
[
    0 => 'a'
    1 => 'b'
    2 => 'c'
    3 => 0  // wrong, should be 10
    4 => 1
]
```

**Comment by Alex-Code** [comment]
This is due to `array_slice` here [ArrayDataProvider#L85](https://github.com/yiisoft/yii2/blob/master/framework/data/ArrayDataProvider.php#L85).
It does support a `preserve_keys` option but I'm not sure if it would affect anything else.

**Comment by mdmunir** [comment]
^ Ya, it going right by add last parameter as true
```php
$models = array_slice($models, $pagination->getOffset(), $pagination->getLimit(), true);
```

**Comment by mdmunir** [comment]
@samdark 
why this is closed? 
I change the code in my local. but you should fix it in your repo. at line what alex-code mension.

**Comment by samdark** [comment]
@mdmunir can you make a pull request?

**Comment by mdmunir** [comment]
I am not. You do @samdark .
I create PR in the pass and it rejected. It make me broken heart until now :grinning: .

**Comment by samdark** [comment]
He-he. Life's harsh.

**Comment by klimov-paul** [comment]
Issue resolved by commit bb4bed4
